experimental/oauther/oauth2.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package oauther
+
+import (
+	"encoding/json"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"golang.org/x/oauth2"
+)
+
+const (
+	oAuth2StateTimeToLive = 300 // seconds
+)
+
+type OAuther interface {
+	GetToken(userID string) (*oauth2.Token, error)
+	GetURL() string
+	Deauth(userID string) error
+}
+
+type oAuther struct {
+	api             plugin.API
+	onConnect       func(userID string, token *oauth2.Token)
+	storePrefix     string
+	pluginURL       string
+	oAuthURL        string
+	connectedString string
+	config          *oauth2.Config
+	logger          logger.Logger
+}
+
+func NewOAuther(
+	r *mux.Router,
+	api plugin.API,
+	pluginURL,
+	oAuthURL,
+	storePrefix,
+	connectedString string,
+	onConnect func(userID string, token *oauth2.Token),
+	oAuthConfig *oauth2.Config,
+	loggerBot logger.Logger,
+) OAuther {
+	o := &oAuther{
+		api:             api,
+		onConnect:       onConnect,
+		storePrefix:     storePrefix,
+		pluginURL:       pluginURL,
+		oAuthURL:        oAuthURL,
+		config:          oAuthConfig,
+		connectedString: connectedString,
+		logger:          loggerBot,
+	}
+
+	o.config.RedirectURL = pluginURL + oAuthURL + "/complete"
+
+	oauth2Router := r.PathPrefix(oAuthURL).Subrouter()
+	oauth2Router.HandleFunc("/connect", o.oauth2Connect).Methods("GET")
+	oauth2Router.HandleFunc("/complete", o.oauth2Complete).Methods("GET")
+
+	return o
+}
+
+func (o *oAuther) GetURL() string {
+	return o.pluginURL + o.oAuthURL + "/connect"
+}
+
+func (o *oAuther) GetToken(userID string) (*oauth2.Token, error) {
+	rawToken, appErr := o.api.KVGet(o.getTokenKey(userID))
```

Comment:
Would this be an opportunity to accept the plugin-api-wrapper client instead, so we don't have to re-implement JSON serialization?